src/json-validator.cpp

```diff
@@ -783,38 +829,49 @@ class object : public schema
 		// for each property in instance
 		for (auto &p : instance.items()) {
 			if (propertyNames_)
-				propertyNames_->validate(ptr, p.key(), e);
+				propertyNames_->validate(ptr, p.key(), patch, e);
 
 			bool a_prop_or_pattern_matched = false;
 			auto schema_p = properties_.find(p.key());
 			// check if it is in "properties"
 			if (schema_p != properties_.end()) {
 				a_prop_or_pattern_matched = true;
-				schema_p->second->validate(ptr / p.key(), p.value(), e);
+				schema_p->second->validate(ptr / p.key(), p.value(), patch, e);
 			}
 
 #ifndef NO_STD_REGEX
 			// check all matching patternProperties
 			for (auto &schema_pp : patternProperties_)
 				if (REGEX_NAMESPACE::regex_search(p.key(), schema_pp.first)) {
 					a_prop_or_pattern_matched = true;
-					schema_pp.second->validate(ptr / p.key(), p.value(), e);
+					schema_pp.second->validate(ptr / p.key(), p.value(), patch, e);
 				}
 #endif
 
 			// check additionalProperties as a last resort
 			if (!a_prop_or_pattern_matched && additionalProperties_) {
 				first_error_handler additional_prop_err;
-				additionalProperties_->validate(ptr / p.key(), p.value(), additional_prop_err);
+				additionalProperties_->validate(ptr / p.key(), p.value(), patch, additional_prop_err);
 				if (additional_prop_err)
 					e.error(ptr, instance, "validation failed for additional property '" + p.key() + "': " + additional_prop_err.message_);
 			}
 		}
 
+		// reverse search
+		for (auto const &prop : properties_) {
+			const auto finding = instance.find(prop.first);
+			if (instance.end() == finding) { // if the prop is not in the instance
+				const auto &defaultValue = prop.second->defaultValue(ptr, instance, e);
+				if (!defaultValue.empty()) { // if default value is available
+					patch.add((ptr / prop.first), defaultValue);
+				}
+			}
+		}
+
```

Comment:
At first, years ago when I started the validator, I was wondering wether it is a not too heavy to carry around the json_pointer and updating it. Now I see that this helped to code your feature. Great.

Reply:
Yes, that helped a lot.